File: packages/pigeon/platform_tests/android_unit_tests/android/app/src/test/java/com/example/android_unit_tests/NullFieldsTest.java
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.example.android_unit_tests;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import org.junit.Test;

public class NullFieldsTest {
  @Test
  public void builderWithValues() {
    NullFields.NullFieldsSearchRequest request =
        new NullFields.NullFieldsSearchRequest.Builder()
            .setQuery("hello")
            .setIdentifier(1L)
            .build();

    NullFields.NullFieldsSearchReply reply =
        new NullFields.NullFieldsSearchReply.Builder()
            .setResult("result")
            .setError("error")
            .setIndices(Arrays.asList(1L, 2L, 3L))
            .setRequest(request)
            .setType(NullFields.NullFieldsSearchReplyType.success)
            .build();

    assertEquals(reply.getResult(), "result");
    assertEquals(reply.getError(), "error");
    assertEquals(reply.getIndices(), Arrays.asList(1L, 2L, 3L));
    assertEquals(reply.getRequest().getQuery(), "hello");
    assertEquals(reply.getType(), NullFields.NullFieldsSearchReplyType.success);
  }

  @Test
  public void builderRequestWithNulls() {
    NullFields.NullFieldsSearchRequest request =
        new NullFields.NullFieldsSearchRequest.Builder().setQuery(null).setIdentifier(1L).build();
  }

  @Test
  public void builderReplyWithNulls() {
    NullFields.NullFieldsSearchReply reply =
        new NullFields.NullFieldsSearchReply.Builder()
            .setResult(null)
            .setError(null)
            .setIndices(null)
            .setRequest(null)
            .setType(null)
            .build();

    assertNull(reply.getResult());
    assertNull(reply.getError());
    assertNull(reply.getIndices());
    assertNull(reply.getRequest());
    assertNull(reply.getType());
  }

  @Test
  public void requestFromMapWithValues() {
    HashMap<String, Object> map = new HashMap<>();
    map.put("query", "hello");
    map.put("identifier", 1L);

    NullFields.NullFieldsSearchRequest request = NullFields.NullFieldsSearchRequest.fromMap(map);
    assertEquals(request.getQuery(), "hello");
  }

  @Test
  public void requestFromMapWithNulls() {
    HashMap<String, Object> map = new HashMap<>();
    map.put("query", null);
    map.put("identifier", 1L);

    NullFields.NullFieldsSearchRequest request = NullFields.NullFieldsSearchRequest.fromMap(map);
    assertNull(request.getQuery());
  }

  @Test
  public void replyFromMapWithValues() {
    HashMap<String, Object> requestMap = new HashMap<>();
    requestMap.put("query", "hello");
    requestMap.put("identifier", 1L);

    HashMap<String, Object> map = new HashMap<>();
    map.put("result", "result");
    map.put("error", "error");
    map.put("indices", Arrays.asList(1L, 2L, 3L));
    map.put("request", requestMap);
    map.put("type", NullFields.NullFieldsSearchReplyType.success.ordinal());

    NullFields.NullFieldsSearchReply reply = NullFields.NullFieldsSearchReply.fromMap(map);
    assertEquals(reply.getResult(), "result");
    assertEquals(reply.getError(), "error");
    assertEquals(reply.getIndices(), Arrays.asList(1L, 2L, 3L));
    assertEquals(reply.getRequest().getQuery(), "hello");
    assertEquals(reply.getType(), NullFields.NullFieldsSearchReplyType.success);
  }

  @Test
  public void replyFromMapWithNulls() {
    HashMap<String, Object> map = new HashMap<>();
    map.put("result", null);
    map.put("error", null);
    map.put("indices", null);
    map.put("request", null);
    map.put("type", null);

    NullFields.NullFieldsSearchReply reply = NullFields.NullFieldsSearchReply.fromMap(map);
    assertNull(reply.getResult());
    assertNull(reply.getError());
    assertNull(reply.getIndices());
    assertNull(reply.getRequest());
    assertNull(reply.getType());
  }

  @Test
  public void requestToMapWithValues() {
    NullFields.NullFieldsSearchRequest request =
        new NullFields.NullFieldsSearchRequest.Builder()
            .setQuery("hello")
            .setIdentifier(1L)
            .build();

    Map<String, Object> map = request.toMap();
    assertEquals(map.get("query"), "hello");
  }

  @Test
  public void requestToMapWithNulls() {
    NullFields.NullFieldsSearchRequest request =
        new NullFields.NullFieldsSearchRequest.Builder().setQuery(null).setIdentifier(1L).build();

    Map<String, Object> map = request.toMap();
    assertNull(map.get("query"));
  }

  @Test
  public void replyToMapWithValues() {
    NullFields.NullFieldsSearchReply reply =
        new NullFields.NullFieldsSearchReply.Builder()
            .setResult("result")
            .setError("error")
            .setIndices(Arrays.asList(1L, 2L, 3L))
            .setRequest(
                new NullFields.NullFieldsSearchRequest.Builder()
                    .setQuery("hello")
                    .setIdentifier(1L)
                    .build())
            .setType(NullFields.NullFieldsSearchReplyType.success)
            .build();

    Map<String, Object> map = reply.toMap();
    assertEquals(map.get("result"), "result");
    assertEquals(map.get("error"), "error");
    assertEquals(map.get("indices"), Arrays.asList(1L, 2L, 3L));
    assertEquals(map.get("request"), reply.getRequest().toMap());
    assertEquals(map.get("type"), NullFields.NullFieldsSearchReplyType.success.ordinal());
  }

  @Test
  public void replyToMapWithNulls() {
    NullFields.NullFieldsSearchReply reply =
        new NullFields.NullFieldsSearchReply.Builder()
            .setResult(null)
            .setError(null)
            .setIndices(null)
            .setRequest(null)
            .setType(null)
            .build();

    Map<String, Object> map = reply.toMap();
    assertNull(map.get("result"));
    assertNull(map.get("error"));
    assertNull(map.get("indices"));
    assertNull(map.get("request"));
    assertNull(map.get("type"));
  }
}

File: packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/NullFieldsTest.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import <Flutter/Flutter.h>
#import <XCTest/XCTest.h>
#import "EchoMessenger.h"
#import "NullFields.gen.h"

///////////////////////////////////////////////////////////////////////////////////////////
@interface NullFieldsSearchRequest ()
+ (NullFieldsSearchRequest *)fromMap:(NSDictionary *)dict;
- (NSDictionary *)toMap;
@end

///////////////////////////////////////////////////////////////////////////////////////////
@interface NullFieldsSearchReply ()
+ (NullFieldsSearchReply *)fromMap:(NSDictionary *)dict;
- (NSDictionary *)toMap;
@end

///////////////////////////////////////////////////////////////////////////////////////////
@interface NullFieldsTest : XCTestCase
@end

///////////////////////////////////////////////////////////////////////////////////////////
@implementation NullFieldsTest

- (void)testMakeWithValues {
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest makeWithQuery:@"hello" identifier:@1];

  NullFieldsSearchReply *reply =
      [NullFieldsSearchReply makeWithResult:@"result"
                                      error:@"error"
                                    indices:@[ @1, @2, @3 ]
                                    request:request
                                       type:NullFieldsSearchReplyTypeSuccess];

  NSArray *indices = @[ @1, @2, @3 ];
  XCTAssertEqualObjects(@"result", reply.result);
  XCTAssertEqualObjects(@"error", reply.error);
  XCTAssertEqualObjects(indices, reply.indices);
  XCTAssertEqualObjects(@"hello", reply.request.query);
  XCTAssertEqual(NullFieldsSearchReplyTypeSuccess, reply.type);
}

- (void)testMakeRequestWithNulls {
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest makeWithQuery:nil identifier:@1];
  XCTAssertNil(request.query);
}

- (void)testMakeReplyWithNulls {
  NullFieldsSearchReply *reply =
      [NullFieldsSearchReply makeWithResult:nil
                                      error:nil
                                    indices:nil
                                    request:nil
                                       type:NullFieldsSearchReplyTypeSuccess];
  XCTAssertNil(reply.result);
  XCTAssertNil(reply.error);
  XCTAssertNil(reply.indices);
  XCTAssertNil(reply.request);
  XCTAssertEqual(NullFieldsSearchReplyTypeSuccess, reply.type);
}

- (void)testRequestFromMapWithValues {
  NSDictionary *map = @{
    @"query" : @"hello",
    @"identifier" : @1,
  };
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest fromMap:map];
  XCTAssertEqualObjects(@"hello", request.query);
}

- (void)testRequestFromMapWithNulls {
  NSDictionary *map = @{
    @"query" : [NSNull null],
    @"identifier" : @1,
  };
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest fromMap:map];
  XCTAssertNil(request.query);
}

- (void)testReplyFromMapWithValues {
  NSDictionary *map = @{
    @"result" : @"result",
    @"error" : @"error",
    @"indices" : @[ @1, @2, @3 ],
    @"request" : @{
      @"query" : @"hello",
      @"identifier" : @1,
    },
    @"type" : @0,
  };

  NSArray *indices = @[ @1, @2, @3 ];
  NullFieldsSearchReply *reply = [NullFieldsSearchReply fromMap:map];
  XCTAssertEqualObjects(@"result", reply.result);
  XCTAssertEqualObjects(@"error", reply.error);
  XCTAssertEqualObjects(indices, reply.indices);
  XCTAssertEqualObjects(@"hello", reply.request.query);
  XCTAssertEqual(NullFieldsSearchReplyTypeSuccess, reply.type);
}

- (void)testReplyFromMapWithNulls {
  NSDictionary *map = @{
    @"result" : [NSNull null],
    @"error" : [NSNull null],
    @"indices" : [NSNull null],
    @"request" : [NSNull null],
    @"type" : [NSNull null],
  };
  NullFieldsSearchReply *reply = [NullFieldsSearchReply fromMap:map];
  XCTAssertNil(reply.result);
  XCTAssertNil(reply.error);
  XCTAssertNil(reply.indices);
  XCTAssertNil(reply.request.query);
  XCTAssertEqual(NullFieldsSearchReplyTypeSuccess, reply.type);
}

- (void)testRequestToMapWithValuess {
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest makeWithQuery:@"hello" identifier:@1];
  NSDictionary *dict = [request toMap];
  XCTAssertEqual(@"hello", dict[@"query"]);
}

- (void)testRequestToMapWithNulls {
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest makeWithQuery:nil identifier:@1];
  NSDictionary *dict = [request toMap];
  XCTAssertEqual([NSNull null], dict[@"query"]);
}

- (void)testReplyToMapWithValuess {
  NullFieldsSearchReply *reply = [NullFieldsSearchReply
      makeWithResult:@"result"
               error:@"error"
             indices:@[ @1, @2, @3 ]
             request:[NullFieldsSearchRequest makeWithQuery:@"hello" identifier:@1]
                type:NullFieldsSearchReplyTypeSuccess];
  NSDictionary *dict = [reply toMap];
  NSArray *indices = @[ @1, @2, @3 ];
  XCTAssertEqualObjects(@"result", dict[@"result"]);
  XCTAssertEqualObjects(@"error", dict[@"error"]);
  XCTAssertEqualObjects(indices, dict[@"indices"]);
  XCTAssertEqualObjects(@"hello", dict[@"request"][@"query"]);
  XCTAssertEqualObjects(@0, dict[@"type"]);
}

- (void)testReplyToMapWithNulls {
  NullFieldsSearchReply *reply =
      [NullFieldsSearchReply makeWithResult:nil
                                      error:nil
                                    indices:nil
                                    request:nil
                                       type:NullFieldsSearchReplyTypeSuccess];
  NSDictionary *dict = [reply toMap];
  XCTAssertEqual([NSNull null], dict[@"result"]);
  XCTAssertEqual([NSNull null], dict[@"error"]);
  XCTAssertEqual([NSNull null], dict[@"indices"]);
  XCTAssertEqual([NSNull null], dict[@"request"]);
}

@end
